File: 2023/fetch_input.rb
require "net/http"
require "uri"
require "FileUtils"
session_cookie = ENV["AOC_SESSION_COOKIE"]

if session_cookie.nil?
	raise "Please set AOC_SESSION_COOKIE as an environment variable: export AOC_SESSION_COOKIE=<INSERT-COOKIE>"
end

def fetch_input(year, day, session_cookie)
	year = year.to_i
	day = day.to_i
	raise "Year must be a 4-digit integer greater than 2000" unless year > 2000 and year < 9999
	raise "Day must be a valid integer between 1 and 31" unless (1..31).include?(day.to_i)

	uri = URI.parse("https://adventofcode.com/#{year}/day/#{day}/input")
	http = Net::HTTP.new(uri.host, uri.port)
	http.use_ssl = true
	headers = {
		"Cookie" => "session=#{session_cookie}"
	}
	request = Net::HTTP::Get.new(uri.request_uri, headers)
	response = http.request(request)
	if response.is_a?(Net::HTTPSuccess)
		response.body.strip
	else
		raise "Failed to fetch input: #{response.code} #{response.message}"
	end
end

if __FILE__ == $0
  year = ARGV[0].to_i
  day = ARGV[1].to_i

  # Validate year and day inputs
  if year.nil? || day.nil?
    puts "Usage: ruby fetch_input.rb YEAR DAY"
    exit 1
  end

  input = fetch_input(year, day, session_cookie)
  day_string = (1..9).include?(day) ? "0#{day.to_s}" : day.to_s

 	directory_filename = "#{day_string}"
  input_filename = "#{day_string}/input.txt"

	FileUtils.mkdir_p(directory_filename)
  File.write(input_filename, input)

  solution_filename = "#{day_string}/solution.rb"
  unless File.exist?(solution_filename)
	  FileUtils.cp("template.rb", solution_filename)
  end
end

File: 2024/rust/src/bin/fetch_input.rs
use reqwest::header::COOKIE;
use std::env;
use std::fs;
use std::path::PathBuf;

fn find_project_root() -> Result<PathBuf, Box<dyn std::error::Error>> {
    let mut current_path = env::current_exe()?.parent().unwrap().to_path_buf();

    while !current_path.join("Cargo.toml").exists() {
        if !current_path.pop() {
            return Err("Could not find the project root (Cargo.toml)".into());
        }
    }
    Ok(current_path)
}


#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let session_cookie = env::var("AOC_SESSION_COOKIE").expect("Please set AOC_SESSION_COOKIE as an environment variable.");

    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: fetch_input YEAR DAY");
        return Ok(());
    }

    let year: i32 = args[1].parse().expect("Year must be a valid 4-digit integer.");
    let day: i32 = args[2].parse().expect("Day must be a valid integer between 1 and 31.");

    if year < 2000 || year > 9999 {
        panic!("Year must be a 4-digit integer greater than 2000.");
    }
    if day < 1 || day > 31 {
        panic!("Day must be a valid integer between 1 and 31.");
    }

    let day_string = format!("{:02}", day);

    let url = format!("https://adventofcode.com/{}/day/{}/input", year, day);
    let client = reqwest::Client::new();
    let response = client
        .get(&url)
        .header(COOKIE, format!("session={}", session_cookie))
        .send()
        .await?;

    if !response.status().is_success() {
        panic!(
            "Failed to fetch input: {} {}",
            response.status(),
            response.text().await?
        );
    }

    let input = response.text().await?;

    let project_root = find_project_root()?;
    let directory = project_root.join("src/bin");
    let input_file = directory.join(format!("input/{}.txt", day_string));

    fs::write(&input_file, input)?;

    let solution_file = directory.join(format!("{}.rs", day));
    if !solution_file.exists() {
        let template = "// Write your solution here\nfn main() {}";
        fs::write(&solution_file, template)?;
    }

    println!("Input saved to {}", input_file.display());
    Ok(())
}
